File: src/imagebuilder/flyingtux/pkg/alpine.py
from base import BasePkg
from os.path import basename
import subprocess
from metux.util import log

class AlpinePkg(BasePkg):
    def __init__(self, param, jail):
        BasePkg.__init__(self, param, jail)
        self.my_repos_text = ''
        self.apk_cmd = 'apk'
        self.check_mandatory(['arch'])

    """[private] call the package manager"""
    def __call(self, cmd, args):
        self.jail.call_cmd([self.apk_cmd, cmd, ('--root=%s' % self.jail['pkg-sysroot'])] + args)
        return True

    """prepare the sysroot image"""
    def prepare(self):
        BasePkg.prepare(self)
        self.img_create_dirs(['etc/apk/cache',
                              'etc/apk/keys',
                              'var/log',
                              'lib/apk/db'])
        self.img_write_file('/etc/apk/arch', self['arch'])
        self.img_write_file('/etc/apk/world', '')

    """add repositories"""
    def add_repos(self, repos):
        if repos is None:
            return False

        # fixme: add sanity checks for downloaded image specs
        for x in repos['urls']:
            self.my_repos_text = self.my_repos_text + "%s\n" % x
        self.img_write_file('/etc/apk/repositories', self.my_repos_text)

        for x in repos['keys']:
            self.img_copy_file(x, '/etc/apk/keys/%s' % basename(x))

    """add/install packages"""
    def add_packages(self, pkgs):
        return self.__call('add', pkgs)

    def finish(self):
        return BasePkg.finish(self)

File: internal/pkg/alpine.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package pkg

import (
	"path/filepath"

	"github.com/metux/flyingtux/internal/buildjail"
	"github.com/metux/flyingtux/internal/util"
)

const apkCmd = "apk"

var alpineInitDirs = []string{
	"etc/apk/cache",
	"etc/apk/keys",
	"var/log",
	"lib/apk/db",
}

// alpine is the APK package manager (port of AlpinePkg).
type alpine struct {
	base
	arch      string
	reposText string
}

func newAlpine(arch string, jail buildjail.BuildJail) (Pkg, error) {
	if arch == "" {
		return nil, util.ConfigError("alpine pkg: missing mandatory field 'arch'")
	}
	return &alpine{
		base: base{jail: jail, initDirs: alpineInitDirs},
		arch: arch,
	}, nil
}

// call invokes apk against the sysroot root (private __call).
func (a *alpine) call(cmd string, args []string) int {
	full := append([]string{apkCmd, cmd, "--root=" + a.jail.SysrootFn("")}, args...)
	return a.jail.CallCmd(full)
}

func (a *alpine) Prepare() error {
	if err := a.prepareBase(); err != nil {
		return err
	}
	if err := a.jail.SysrootWriteFile("/etc/apk/arch", a.arch); err != nil {
		return err
	}
	return a.jail.SysrootWriteFile("/etc/apk/world", "")
}

func (a *alpine) AddRepos(repos *Repos) error {
	if repos == nil {
		return nil
	}
	for _, x := range repos.URLs {
		a.reposText += x + "\n"
	}
	if err := a.jail.SysrootWriteFile("/etc/apk/repositories", a.reposText); err != nil {
		return err
	}
	for _, x := range repos.Keys {
		if err := a.jail.SysrootCopyFile(x, "/etc/apk/keys/"+filepath.Base(x), "", ""); err != nil {
			return err
		}
	}
	return nil
}

func (a *alpine) AddPackages(pkgs []string) error {
	if len(pkgs) == 0 {
		return nil
	}
	if code := a.call("add", pkgs); code != 0 {
		return util.Error("apk add failed (exit %d)", code)
	}
	return nil
}
